hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java

```diff
@@ -90,14 +99,26 @@ private SqlPage fetchInternal(
     ) {
         ResultIterator<SqlRow> iterator = clientCursor.getIterator();
 
-        List<List<Data>> page = new ArrayList<>(cursorBufferSize);
-        boolean last = fetchPage(iterator, page, cursorBufferSize, serializationService, isFirstPage);
+        try {
+            List<List<Data>> page = new ArrayList<>(cursorBufferSize);
+            boolean last = fetchPage(iterator, page, cursorBufferSize, serializationService, isFirstPage);
 
-        if (last) {
-            deleteClientCursor(clientCursor);
-        }
+            return new SqlPage(page, last);
+        } catch (HazelcastSqlException e) {
+            // We use public API to extract results from the cursor. The cursor may throw HazelcastSqlException only. When
+            // it happens, the cursor is already closed with the error, so we just re-throw.
+            throw e;
+        } catch (Exception e) {
+            // Any other exception indicates that something has happened outside of the internal query state. For example,
+            // we may fail to serialize a specific column value to Data. We have to close the cursor in this case.
+            AbstractSqlResult result = clientCursor.getSqlResult();
+
+            QueryException error = QueryException.error(e.getMessage(), e);
```

Comment:
The exception class is available in the stack trace, there is not much need for duplicating it. Instead, I added a specific error message to this line of code, so that it is easier to identify where it happened on the client-side (where the stack trace is not available)